src/solver.rs

```diff
@@ -38,6 +44,165 @@ pub struct CandidateBuckets {
     pub count: usize,
 }
 
+/// Build a stable key for a full solution (bindings in **pattern order**).
+/// Prefer the whole word if present (WORD_SENTINEL). Fall back to sorted (var,val) pairs.
+fn solution_key(solution: &[Bindings]) -> u64 {
+    let mut hasher = DefaultHasher::new();
+
+    for b in solution {
+        // Try whole-word first (fast + canonical)
+        if let Some(w) = b.get_word() {
+            w.hash(&mut hasher);
+        } else {
+            // this should never happen
+            /*
+            // Fall back: hash all (var,val) pairs sorted by var
+            let mut pairs: Vec<(char, String)> =
+                b.iter().map(|(k, v)| (*k, v.clone())).collect();
+            pairs.sort_unstable_by_key(|(k, _)| *k);
+            for (k, v) in pairs {
+                k.hash(&mut hasher);
+                v.hash(&mut hasher);
+            }
+            */
+        }
+        // Separator between patterns to avoid ambiguity like ["ab","c"] vs ["a","bc"]
+        0xFFFFu16.hash(&mut hasher);
+    }
+
+    hasher.finish()
+}
+
+/// Simple helper to enforce a wall-clock time limit.
+///
+/// Usage:
+///   let budget = TimeBudget::new(Duration::from_secs(30));
+///   while !budget.expired() {
+///       // do some work
+///   }
+///
+/// You can also query how much time is left (`remaining()`).
+struct TimeBudget {
+    start: Instant,   // when the budget began
+    limit: Duration,  // maximum allowed elapsed time
+}
+
+impl TimeBudget {
+    /// Create a new budget that lasts for `limit` (e.g., 30 seconds).
+    fn new(limit: Duration) -> Self {
+        Self { start: Instant::now(), limit }
+    }
+
+    /// Returns true if the allowed time has fully elapsed.
+    fn expired(&self) -> bool {
+        self.start.elapsed() >= self.limit
+    }
+
+    // Returns the remaining time before expiration, or zero if the budget is already used up.
+    // Unused for now but it may be useful later
+    // fn remaining(&self) -> Duration {self.limit.saturating_sub(self.start.elapsed())}
+}
+
+
+/// Build the deterministic lookup key for a binding given the pattern's lookup vars.
+fn lookup_key_for_binding(binding: &Bindings, keys_opt: &Option<HashSet<char>>) -> LookupKey {
+    match keys_opt {
+        None => None, // pattern has no lookup constraints
+        Some(keys) => {
+            // Collect (var, value) only for vars present in this binding.
+            let mut pairs: Vec<(char, String)> = Vec::with_capacity(keys.len());
+            for &var in keys {
+                if let Some(val) = binding.get(var) {
+                    pairs.push((var, val.clone()));
+                } else {
+                    // Missing a required key → this binding is unusable for join
+                    pairs.clear();
+                    break;
+                }
+            }
+            if pairs.is_empty() && !keys.is_empty() {
```

Comment:
why `!keys.is_empty()`? if `keys_is_empty()`, won't `pairs` be empty, so won't the returns be more or less similar? (currently, `keys_is_empty()` would seem to lead to returning an empty `Vec` with capacity 0 and the positive if case (123–127) returns an empty `Vec` with _default_ capacity